tests/nn/data_parallel/test_fsdp_optimizer_utils.py

```diff
@@ -0,0 +1,123 @@
+from fairscale.utils.testing import objects_are_equal
+
+from .test_fsdp import (
+    DistributedTest,
+    DummyProcessGroup,
+    NestedWrappedModule,
+    TransformerWithSharedParams,
+    rename_test,
+    spawn_and_init,
+)
+
+
+def first_tensor_shape(dct):
+    for k, v in dct.items():
+        if torch.is_tensor(v):
+            return v.numel()
+    return 0
+
+
+def assert_equal(a, b):
+    assert a == b, f"{a} != {b}"
+
+
+class TestOptimizerUtils(DistributedTest):
+    @parameterized.expand(
+        [[functools.partial(SGD, momentum=0.9), True], [SGD, False], [Adam, False], [Adadelta, True]],
+        name_func=rename_test,
+    )
+    def test_consolidate_optimizer(self, optim_fn, transformer):
+        config = {"mixed_precision": True, "flatten_parameters": True}
+        test_fn = functools.partial(
+            self._test_consolidated_optimizer, config, optim_fn=optim_fn, transformer=transformer
+        )
+
+        spawn_and_init(test_fn, world_sizes=[min(torch.cuda.device_count(), 4)])
+
+    @classmethod
+    def _test_consolidated_optimizer(self, config, rank, group, optim_fn=torch.optim.SGD, transformer=False):
+        """FSDP.gather_full_optim_state_dict() should return something very similar to optimizer.state_dict()"""
+        # Establish reference behavior.
+
+        if transformer:
+            fsdp = self.get_wrapped_model(group, config=config).cuda()
+            unwrapped_model = TransformerWithSharedParams(group).cuda()
+        else:
+            fsdp = FullyShardedDataParallel(NestedWrappedModule(group, wrapper_config=config), group, **config).cuda()
+            unwrapped_model = NestedWrappedModule(group, wrapper_config=None).cuda()
+
+        try:
+            fsdp_optim = optim_fn(fsdp.parameters(), lr=0.01,)
+            optim_unwrapped = optim_fn(unwrapped_model.parameters(), lr=0.01)
+        except TypeError:  # AdaScale
+            fsdp_optim = optim_fn(fsdp.parameters())
+            optim_unwrapped = optim_fn(unwrapped_model.parameters())
+
+        fsdp_optim.zero_grad()
+        optim_unwrapped.zero_grad()
+
+        x = fsdp.module.get_input(torch.device("cuda"))
+        output = fsdp(*x)
+        loss = fsdp.module.get_loss(x, output).to("cuda")
+        fsdp.module.run_backward(loss)
+        fsdp_optim.step()
+
+        output = unwrapped_model(*x)
+        loss = unwrapped_model.get_loss(x, output)
+        unwrapped_model.run_backward(loss)
+        optim_unwrapped.step()
+        unwrapped_sd = optim_unwrapped.state_dict()
+
+        # first_key = unwrapped_sd['state'][0].keys()
+        tstart = time()
+        sd = fsdp.gather_full_optim_state_dict(fsdp_optim, recipient_rank=0)
+        duration = time() - tstart
+        # Switching from fairscale.optim.utils.broadcast_object to torch.broadcast_object_list will cause this to raise
+        assert duration < fsdp.world_size, f"gather optim state took {duration} seconds, suspect change in _consolidate"
```

Comment:
this is interesting. thanks for the comment. what the usual value for `duration`? I am surprised that it is somehow connected with `world_size`, which is not in the unit of seconds even.

Reply:
It takes longer to gather from 8 nodes than 4 than 2.
This actually takes 4 ms, but I accidentally regressed it during development and caused it to take 8 seconds for world size 2, 13 for world size 4. 
Now that it's fixed I want to prevent it happening again, agreed that the units are arbitrary.